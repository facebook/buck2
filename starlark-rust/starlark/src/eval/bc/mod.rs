/*
 * Copyright 2019 The Starlark in Rust Authors.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//! Bytecode interpreter.

pub(crate) mod addr;
pub(crate) mod bytecode;
pub(crate) mod call;
pub(crate) mod compiler;
pub(crate) mod definitely_assigned;
pub(crate) mod frame;
pub(crate) mod if_debug;
pub(crate) mod instr;
pub(crate) mod instr_arg;
pub(crate) mod instr_impl;
pub(crate) mod instrs;
pub(crate) mod native_function;
pub(crate) mod opcode;
pub(crate) mod repr;
pub(crate) mod slow_arg;
pub(crate) mod stack_ptr;
pub(crate) mod stack_values;
pub(crate) mod writer;
