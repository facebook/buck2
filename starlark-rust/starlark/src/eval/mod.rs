/*
 * Copyright 2018 The Starlark in Rust Authors.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//! Evaluate some code, typically done by creating an [`Evaluator`], then calling
//! [`eval_module`](Evaluator::eval_module).

pub(crate) mod bc;
pub(crate) mod compiler;
pub(crate) mod runtime;

use std::collections::HashMap;
use std::mem;
use std::time::Instant;

use dupe::Dupe;
pub use runtime::arguments::Arguments;
pub use runtime::before_stmt::BeforeStmtFuncDyn;
pub use runtime::call_stack::CallStack;
pub use runtime::evaluator::Evaluator;
pub use runtime::file_loader::FileLoader;
pub use runtime::file_loader::ReturnFileLoader;
pub use runtime::params::ParametersParser;
pub use runtime::params::ParametersSpec;
pub use runtime::params::ParametersSpecBuilder;
pub use runtime::profile::data::ProfileData;
pub use runtime::profile::ProfileMode;

use crate::collections::symbol_map::Symbol;
use crate::docs::DocString;
use crate::environment::Globals;
use crate::eval::compiler::def::DefInfo;
use crate::eval::compiler::scope::ModuleScopes;
use crate::eval::compiler::scope::ScopeId;
use crate::eval::compiler::Compiler;
use crate::eval::compiler::EvalException;
use crate::eval::runtime::arguments::ArgNames;
use crate::eval::runtime::arguments::ArgumentsFull;
use crate::slice_vec_ext::SliceExt;
use crate::syntax::module::AstModule;
use crate::syntax::DialectTypes;
use crate::values::Value;

impl<'v, 'a> Evaluator<'v, 'a> {
    /// Evaluate an [`AstModule`] with this [`Evaluator`], modifying the in-scope
    /// [`Module`](crate::environment::Module) as appropriate.
    pub fn eval_module(&mut self, ast: AstModule, globals: &Globals) -> anyhow::Result<Value<'v>> {
        let start = Instant::now();

        let AstModule {
            codemap,
            statement,
            dialect,
        } = ast;

        let codemap = self
            .module_env
            .frozen_heap()
            .alloc_any_display_from_debug(codemap.dupe());

        let globals = self
            .module_env
            .frozen_heap()
            .alloc_any_display_from_type_name(globals.dupe());

        if let Some(docstring) = DocString::extract_raw_starlark_docstring(&statement) {
            self.module_env.set_docstring(docstring)
        }

        let (statement, mut scope) = ModuleScopes::enter_module(
            self.module_env.mutable_names(),
            self.module_env.frozen_heap(),
            &HashMap::new(),
            statement,
            globals,
            codemap,
            &dialect,
        );

        // We want to grab the first error only, with ownership, so drop all but the first
        scope.errors.truncate(1);
        if let Some(e) = scope.errors.pop() {
            // Static errors, reported even if the branch is not hit
            return Err(e.into_anyhow());
        }

        let (module_slots, scope_data) = scope.exit_module();
        let scope_names = scope_data.get_scope(ScopeId::module());
        let local_names = self
            .frozen_heap()
            .alloc_any_slice_display_from_debug(&scope_names.used);

        self.module_env.slots().ensure_slots(module_slots);
        let old_def_info = mem::replace(
            &mut self.module_def_info,
            self.module_env.frozen_heap().alloc_any(DefInfo::for_module(
                codemap,
                local_names,
                self.module_env
                    .frozen_heap()
                    .alloc_any_slice_display_from_debug(&scope_names.parent),
                globals,
            )),
        );

        // Set up the world to allow evaluation (do NOT use ? from now on)

        self.call_stack.push(Value::new_none(), None).unwrap();

        // Evaluation
        let mut compiler = Compiler {
            scope_data,
            locals: Vec::new(),
            globals,
            codemap,
            eval: self,
            check_types: dialect.enable_types == DialectTypes::Enable,
        };

        let res = compiler.eval_module(statement, local_names);

        // Clean up the world, putting everything back
        self.call_stack.pop();

        self.module_def_info = old_def_info;

        self.module_env.add_eval_duration(start.elapsed());

        // Return the result of evaluation
        res.map_err(EvalException::into_anyhow)
    }

    /// Evaluate a function stored in a [`Value`], passing in `positional` and `named` arguments.
    pub fn eval_function(
        &mut self,
        function: Value<'v>,
        positional: &[Value<'v>],
        named: &[(&str, Value<'v>)],
    ) -> anyhow::Result<Value<'v>> {
        let names = named.map(|(s, _)| (Symbol::new(s), self.heap().alloc_str(s)));
        let named = named.map(|x| x.1);
        let params = Arguments(ArgumentsFull {
            pos: positional,
            named: &named,
            names: ArgNames::new(&names),
            args: None,
            kwargs: None,
        });
        // eval_module pushes an "empty" call stack frame. other places expect that first frame to be ignorable, and
        // so we push an empty frame too (otherwise things would ignore this function's own frame).
        self.with_call_stack(Value::new_none(), None, |this| {
            function.invoke(&params, this)
        })
    }
}
