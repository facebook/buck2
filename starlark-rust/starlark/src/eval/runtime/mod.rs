/*
 * Copyright 2019 The Starlark in Rust Authors.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

pub(crate) mod arguments;
pub(crate) mod before_stmt;
pub(crate) mod call_stack;
pub(crate) mod evaluator;
pub(crate) mod file_loader;
pub(crate) mod inlined_frame;
pub(crate) mod params;
pub(crate) mod profile;
pub(crate) mod rust_loc;
pub(crate) mod slots;
pub(crate) mod small_duration;
pub(crate) mod visit_span;
