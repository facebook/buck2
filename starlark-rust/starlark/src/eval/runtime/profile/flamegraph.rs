/*
 * Copyright 2019 The Starlark in Rust Authors.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//! Utility to write files in formats understood by `flamegraph.pl`.

use std::fmt::Write;

use dupe::Dupe;
use starlark_map::small_map::SmallMap;

use crate::util::arc_str::ArcStr;

/// Node in flamegraph tree.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub(crate) struct FlameGraphNode {
    children: SmallMap<ArcStr, FlameGraphNode>,
    value: Option<u64>,
}

/// Profiling data as flame tree.
///
/// Can be written to `flamegraph.pl` format.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub(crate) struct FlameGraphData {
    root: FlameGraphNode,
}

impl FlameGraphNode {
    fn write<'a>(&'a self, writer: &mut FlameGraphWriter, stack: &mut Vec<&'a str>) {
        if let Some(value) = self.value {
            writer.write(stack.iter().copied(), value);
        }
        for (k, v) in self.children.iter() {
            stack.push(k);
            v.write(writer, stack);
            stack.pop().unwrap();
        }
    }

    /// Add value to the node.
    pub(crate) fn add(&mut self, value: u64) {
        match &mut self.value {
            None => self.value = Some(value),
            Some(v) => *v += value,
        }
    }

    pub(crate) fn merge(&mut self, other: &FlameGraphNode) {
        if let Some(value) = other.value {
            self.add(value);
        }

        for (k, v) in &other.children {
            self.child(k.dupe()).merge(v);
        }
    }

    /// Get or create a child node.
    pub(crate) fn child(&mut self, name: ArcStr) -> &mut FlameGraphNode {
        self.children.entry(name).or_default()
    }
}

impl FlameGraphData {
    pub(crate) fn write(&self) -> String {
        let mut writer = FlameGraphWriter::new();
        let mut stack = Vec::new();
        self.root.write(&mut writer, &mut stack);
        assert!(stack.is_empty());
        writer.finish()
    }

    pub(crate) fn root(&mut self) -> &mut FlameGraphNode {
        &mut self.root
    }

    pub(crate) fn merge<'a>(
        graphs: impl IntoIterator<Item = &'a FlameGraphData>,
    ) -> FlameGraphData {
        let mut result = FlameGraphData::default();
        for graph in graphs {
            result.root.merge(&graph.root);
        }
        result
    }
}

pub(crate) struct FlameGraphWriter {
    buf: String,
}

impl FlameGraphWriter {
    pub(crate) fn new() -> FlameGraphWriter {
        FlameGraphWriter { buf: String::new() }
    }

    pub(crate) fn write<'s>(&mut self, key: impl IntoIterator<Item = &'s str>, value: u64) {
        let key = key.into_iter().collect::<Vec<_>>();
        if key.is_empty() {
            writeln!(self.buf, "(unknown) {}", value).unwrap();
        } else {
            writeln!(self.buf, "{} {}", key.join(";"), value).unwrap();
        }
    }

    pub(crate) fn finish(self) -> String {
        self.buf
    }
}

#[cfg(test)]
mod tests {
    use crate::eval::runtime::profile::flamegraph::FlameGraphData;
    use crate::eval::runtime::profile::flamegraph::FlameGraphWriter;

    #[test]
    fn test_flamegraph_writer() {
        let mut writer = FlameGraphWriter::new();
        writer.write(["aa", "bb"], 20);
        assert_eq!("aa;bb 20\n", writer.finish());
    }

    #[test]
    fn test_flamegraph_data() {
        let mut data = FlameGraphData::default();
        data.root().child("a".into()).add(10);
        data.root().child("a".into()).child("b".into()).add(20);
        data.root().child("a".into()).add(30);
        let data = data.write();
        assert_eq!("a 40\na;b 20\n", data);
    }

    #[test]
    fn test_merge() {
        let mut a = FlameGraphData::default();
        a.root().add(10);
        a.root().child("a".into()).add(100);
        a.root().child("b".into()).child("c".into()).add(1000);
        let mut b = FlameGraphData::default();
        b.root().add(20);
        b.root().child("a".into()).add(200);

        let c = FlameGraphData::merge([&a, &b]);

        let mut expected = FlameGraphData::default();
        expected.root().add(30);
        expected.root().child("a".into()).add(300);
        expected
            .root()
            .child("b".into())
            .child("c".into())
            .add(1000);

        assert_eq!(expected, c);
    }
}
