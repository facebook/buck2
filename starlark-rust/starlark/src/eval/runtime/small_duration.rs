/*
 * Copyright 2019 The Starlark in Rust Authors.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::{
    iter::Sum,
    ops::{Add, AddAssign, Div},
    time::Duration,
};

use gazebo::dupe::Dupe;

/// Slightly faster than `Duration`.
#[derive(Copy, Clone, Dupe, Default, Debug)]
pub(crate) struct SmallDuration {
    /// `u64::MAX` nanos is 500 years.
    pub(crate) nanos: u64,
}

impl SmallDuration {
    pub(crate) fn from_duration(duration: Duration) -> SmallDuration {
        SmallDuration {
            nanos: duration.as_nanos() as u64,
        }
    }

    pub(crate) fn to_duration(self) -> Duration {
        Duration::from_nanos(self.nanos)
    }
}

impl AddAssign for SmallDuration {
    fn add_assign(&mut self, other: Self) {
        self.nanos += other.nanos;
    }
}

impl AddAssign<Duration> for SmallDuration {
    fn add_assign(&mut self, other: Duration) {
        self.nanos += other.as_nanos() as u64;
    }
}

impl Add<Duration> for SmallDuration {
    type Output = SmallDuration;

    fn add(self, other: Duration) -> Self::Output {
        SmallDuration {
            nanos: self.nanos + other.as_nanos() as u64,
        }
    }
}

impl Add<SmallDuration> for SmallDuration {
    type Output = SmallDuration;

    fn add(self, other: SmallDuration) -> SmallDuration {
        SmallDuration {
            nanos: self.nanos + other.nanos,
        }
    }
}

impl Div<u64> for SmallDuration {
    type Output = SmallDuration;

    fn div(self, other: u64) -> SmallDuration {
        SmallDuration {
            nanos: self.nanos / other,
        }
    }
}

impl<'a> Sum<&'a SmallDuration> for SmallDuration {
    fn sum<I>(iter: I) -> SmallDuration
    where
        I: Iterator<Item = &'a SmallDuration>,
    {
        iter.fold(SmallDuration::default(), |acc, x| acc + *x)
    }
}
