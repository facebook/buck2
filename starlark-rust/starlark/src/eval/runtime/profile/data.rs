/*
 * Copyright 2019 The Starlark in Rust Authors.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::fs;
use std::path::Path;

use anyhow::Context;
use gazebo::prelude::*;

use crate::eval::runtime::profile::bc::BcPairsProfileData;
use crate::eval::runtime::profile::bc::BcProfileData;
use crate::eval::ProfileMode;
use crate::values::AggregateHeapProfileInfo;

#[derive(Debug, thiserror::Error)]
enum ProfileDataError {
    #[error("Profile data is not consistent with profile mode (internal error)")]
    ProfileDataNotConsistent,
    #[error("Empty profile list cannot be merged")]
    EmptyProfileList,
    #[error("Different profile modes in profile")]
    DifferentProfileModes,
    #[error("Merge of profile data for profile mode `{0}` is not implemented")]
    MergeNotImplemented(ProfileMode),
}

#[derive(Clone, Debug)]
pub(crate) enum ProfileDataImpl {
    Bc(Box<BcProfileData>),
    BcPairs(BcPairsProfileData),
    AggregateHeapProfileInfo(Box<AggregateHeapProfileInfo>),
    Other(String),
}

/// Collected profiling data.
#[derive(Clone, Debug)]
pub struct ProfileData {
    pub(crate) profile_mode: ProfileMode,
    /// Serialized to text (e.g. CSV or flamegraph).
    pub(crate) profile: ProfileDataImpl,
}

impl ProfileData {
    pub(crate) fn new(profile_mode: ProfileMode, profile: String) -> ProfileData {
        ProfileData {
            profile_mode,
            profile: ProfileDataImpl::Other(profile),
        }
    }

    /// Generate a string with profile data (e.g. CSV or flamegraph, depending on profile type).
    pub fn gen(&self) -> anyhow::Result<String> {
        match (&self.profile, &self.profile_mode) {
            (ProfileDataImpl::Other(profile), _) => Ok(profile.clone()),
            (ProfileDataImpl::Bc(bc), _) => Ok(bc.gen_csv()),
            (ProfileDataImpl::BcPairs(bc_pairs), _) => Ok(bc_pairs.gen_csv()),
            (
                ProfileDataImpl::AggregateHeapProfileInfo(profile),
                ProfileMode::HeapFlameRetained | ProfileMode::HeapFlameAllocated,
            ) => Ok(profile.gen_flame_graph()),
            (
                ProfileDataImpl::AggregateHeapProfileInfo(profile),
                ProfileMode::HeapSummaryRetained | ProfileMode::HeapSummaryAllocated,
            ) => Ok(profile.gen_summary_csv()),
            (ProfileDataImpl::AggregateHeapProfileInfo(_), _) => {
                Err(ProfileDataError::ProfileDataNotConsistent.into())
            }
        }
    }

    /// Write to a file.
    pub fn write(&self, path: &Path) -> anyhow::Result<()> {
        fs::write(path, &self.gen()?).with_context(|| {
            format!(
                "write profile `{}` data to `{}`",
                self.profile_mode,
                path.display()
            )
        })?;
        Ok(())
    }

    /// Merge profiles (aggregate).
    pub fn merge<'a>(
        profiles: impl IntoIterator<Item = &'a ProfileData>,
    ) -> anyhow::Result<ProfileData> {
        let profiles = Vec::from_iter(profiles);
        let profile_mode = match profiles.first() {
            None => return Err(ProfileDataError::EmptyProfileList.into()),
            Some(p) => p.profile_mode.dupe(),
        };
        for p in &profiles {
            if p.profile_mode != profile_mode {
                return Err(ProfileDataError::DifferentProfileModes.into());
            }
        }
        let profile = match &profile_mode {
            ProfileMode::Bytecode => {
                let profiles = profiles.try_map(|p| match &p.profile {
                    ProfileDataImpl::Bc(bc) => Ok(&**bc),
                    _ => Err(ProfileDataError::ProfileDataNotConsistent),
                })?;
                let profile = BcProfileData::merge(profiles);
                ProfileDataImpl::Bc(box profile)
            }
            ProfileMode::BytecodePairs => {
                let profiles = profiles.try_map(|p| match &p.profile {
                    ProfileDataImpl::BcPairs(bc_pairs) => Ok(bc_pairs),
                    _ => Err(ProfileDataError::ProfileDataNotConsistent),
                })?;
                let profile = BcPairsProfileData::merge(profiles);
                ProfileDataImpl::BcPairs(profile)
            }
            ProfileMode::HeapSummaryAllocated
            | ProfileMode::HeapSummaryRetained
            | ProfileMode::HeapFlameAllocated
            | ProfileMode::HeapFlameRetained => {
                let profiles = profiles.try_map(|p| match &p.profile {
                    ProfileDataImpl::AggregateHeapProfileInfo(profile) => Ok(&**profile),
                    _ => Err(ProfileDataError::ProfileDataNotConsistent),
                })?;
                let profile = AggregateHeapProfileInfo::merge(profiles);
                ProfileDataImpl::AggregateHeapProfileInfo(box profile)
            }
            profile_mode => {
                return Err(ProfileDataError::MergeNotImplemented(profile_mode.dupe()).into());
            }
        };
        Ok(ProfileData {
            profile_mode,
            profile,
        })
    }
}

#[cfg(test)]
mod tests {
    use gazebo::dupe::Dupe;

    use crate::eval::runtime::profile::bc::BcPairsProfileData;
    use crate::eval::runtime::profile::bc::BcProfileData;
    use crate::eval::runtime::profile::data::ProfileDataImpl;
    use crate::eval::ProfileData;
    use crate::eval::ProfileMode;
    use crate::values::AggregateHeapProfileInfo;

    #[test]
    fn merge_bc() {
        let profile = ProfileData {
            profile_mode: ProfileMode::Bytecode,
            profile: ProfileDataImpl::Bc(box BcProfileData::default()),
        };
        // Smoke.
        ProfileData::merge([&profile, &profile]).unwrap();
    }

    #[test]
    fn merge_bc_pairs() {
        let profile = ProfileData {
            profile_mode: ProfileMode::BytecodePairs,
            profile: ProfileDataImpl::BcPairs(BcPairsProfileData::default()),
        };
        // Smoke.
        ProfileData::merge([&profile, &profile]).unwrap();
    }

    #[test]
    fn merge_aggregated_heap_profile() {
        for profile_mode in [
            ProfileMode::HeapFlameRetained,
            ProfileMode::HeapFlameAllocated,
            ProfileMode::HeapSummaryRetained,
            ProfileMode::HeapSummaryAllocated,
        ] {
            let profile = ProfileData {
                profile_mode: profile_mode.dupe(),
                profile: ProfileDataImpl::AggregateHeapProfileInfo(
                    box AggregateHeapProfileInfo::default(),
                ),
            };
            // Smoke.
            ProfileData::merge([&profile, &profile]).unwrap();
        }
    }
}
