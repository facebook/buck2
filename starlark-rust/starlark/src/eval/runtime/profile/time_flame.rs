/*
 * Copyright 2019 The Starlark in Rust Authors.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::slice;
use std::time::Instant;

use dupe::Dupe;
use starlark_map::StarlarkHasherBuilder;

use crate as starlark;
use crate::eval::runtime::profile::data::ProfileData;
use crate::eval::runtime::profile::data::ProfileDataImpl;
use crate::eval::runtime::profile::flamegraph::FlameGraphData;
use crate::eval::runtime::profile::flamegraph::FlameGraphNode;
use crate::eval::runtime::small_duration::SmallDuration;
use crate::eval::ProfileMode;
use crate::slice_vec_ext::SliceExt;
use crate::values::layout::heap::profile::arc_str::ArcStr;
use crate::values::layout::pointer::RawPointer;
use crate::values::FrozenValue;
use crate::values::Trace;
use crate::values::Tracer;
use crate::values::Value;

#[derive(Debug, thiserror::Error)]
enum FlameProfileError {
    #[error("Flame profile not enabled")]
    NotEnabled,
}

#[derive(Hash, PartialEq, Eq, Clone, Copy, Dupe)]
struct MutableValueId(usize);
#[derive(Hash, PartialEq, Eq, Clone, Copy, Dupe)]
struct FrozenValueId(usize);

/// Index into FlameData.values
#[derive(Hash, PartialEq, Eq, Clone, Copy, Dupe)]
enum ValueId {
    // This struct is two words, can be made one.
    Mutable(MutableValueId),
    Frozen(FrozenValueId),
}

impl ValueId {
    fn lookup<'a, T>(self, mutable: &'a [T], frozen: &'a [T]) -> &'a T {
        match self {
            ValueId::Mutable(x) => &mutable[x.0],
            ValueId::Frozen(x) => &frozen[x.0],
        }
    }
}

/// Bimap between `Value` and `ValueId`.
/// In order to optimise GC (which otherwise quickly becomes O(n^2)) we have to
/// dedupe the values, so store them in `values`, with a fast map to get them in `map`.
/// Whenever we GC, regenerate map.
#[derive(Default)]
struct ValueIndex<'v> {
    /// Map from `MutableValueId` to `Value`.
    mutable_values: Vec<Value<'v>>,
    /// Map from `FrozenValueId` to `Value`.
    frozen_values: Vec<FrozenValue>,
    /// Map from `Value` to `MutableValueId`.
    mutable_map: HashMap<RawPointer, MutableValueId, StarlarkHasherBuilder>,
    /// Map from `Value` to `FrozenValueId`.
    frozen_map: HashMap<RawPointer, FrozenValueId, StarlarkHasherBuilder>,
}

unsafe impl<'v> Trace<'v> for ValueIndex<'v> {
    fn trace(&mut self, tracer: &Tracer<'v>) {
        // We only need to trace mutable values.
        self.mutable_values.trace(tracer);
        // Have to rebuild the map, as its keyed by ValuePtr which changes on GC
        self.mutable_map.clear();
        for (i, x) in self.mutable_values.iter().enumerate() {
            self.mutable_map.insert(x.ptr_value(), MutableValueId(i));
        }
    }
}

impl<'v> ValueIndex<'v> {
    /// Map `Value` to `ValueId`.
    fn index(&mut self, value: Value<'v>) -> ValueId {
        match value.unpack_frozen() {
            Some(frozen) => match self.frozen_map.entry(frozen.ptr_value()) {
                Entry::Occupied(e) => ValueId::Frozen(*e.get()),
                Entry::Vacant(e) => {
                    let res = FrozenValueId(self.frozen_values.len());
                    self.frozen_values.push(frozen);
                    e.insert(res);
                    ValueId::Frozen(res)
                }
            },
            None => match self.mutable_map.entry(value.ptr_value()) {
                Entry::Occupied(e) => ValueId::Mutable(*e.get()),
                Entry::Vacant(e) => {
                    let res = MutableValueId(self.mutable_values.len());
                    self.mutable_values.push(value);
                    e.insert(res);
                    ValueId::Mutable(res)
                }
            },
        }
    }
}

enum Frame {
    /// Entry recorded when we enter a function.
    Push(ValueId),
    /// Entry recorded when we exit a function.
    Pop,
}

#[derive(Trace)]
pub(crate) struct TimeFlameProfile<'v>(
    /// `Some` means enabled.
    Option<Box<FlameData<'v>>>,
);

#[derive(Default, Trace)]
struct FlameData<'v> {
    /// All events in the profile, i.e. function entry or exit with timestamp.
    frames: Vec<(Frame, Instant)>,
    index: ValueIndex<'v>,
}

struct Stacks<'a> {
    name: &'a str,
    time: SmallDuration,
    children: HashMap<ValueId, Stacks<'a>, StarlarkHasherBuilder>,
}

impl<'a> Stacks<'a> {
    fn blank(name: &'a str) -> Self {
        Stacks {
            name,
            time: SmallDuration::default(),
            children: HashMap::with_hasher(StarlarkHasherBuilder::default()),
        }
    }

    fn new(
        mutable_names: &'a [String],
        frozen_names: &'a [String],
        frames: &[(Frame, Instant)],
    ) -> Self {
        let mut res = Stacks::blank("root");
        let Some(mut last_time) = frames.first().map(|x| x.1) else {
            return res;
        };
        res.add(
            mutable_names,
            frozen_names,
            &mut frames.iter(),
            &mut last_time,
        );
        res
    }

    fn add(
        &mut self,
        mutable_names: &'a [String],
        frozen_names: &'a [String],
        frames: &mut slice::Iter<(Frame, Instant)>,
        last_time: &mut Instant,
    ) {
        while let Some((frame, time)) = frames.next() {
            self.time += time.duration_since(*last_time);
            *last_time = *time;
            match frame {
                Frame::Pop => return,
                Frame::Push(i) => match self.children.entry(*i) {
                    Entry::Occupied(mut e) => {
                        e.get_mut()
                            .add(mutable_names, frozen_names, frames, last_time)
                    }
                    Entry::Vacant(e) => e
                        .insert(Stacks::blank(
                            i.lookup(mutable_names, frozen_names).as_str(),
                        ))
                        .add(mutable_names, frozen_names, frames, last_time),
                },
            }
        }
    }

    fn render_with_buffer(&self, node: &mut FlameGraphNode) {
        let node = node.child(ArcStr::from(self.name));
        let count = self.time.to_duration().as_millis();
        if count > 0 {
            node.add(count as u64);
        }
        for x in self.children.values() {
            x.render_with_buffer(node);
        }
    }

    fn render(&self) -> FlameGraphData {
        let mut data = FlameGraphData::default();
        self.render_with_buffer(data.root());
        data
    }
}

impl<'v> TimeFlameProfile<'v> {
    pub(crate) fn new() -> Self {
        Self(None)
    }

    pub(crate) fn enable(&mut self) {
        self.0 = Some(Box::default());
    }

    pub(crate) fn record_call_enter(&mut self, function: Value<'v>) {
        if let Some(x) = &mut self.0 {
            let ind = x.index.index(function);
            x.frames.push((Frame::Push(ind), Instant::now()))
        }
    }

    pub(crate) fn record_call_exit(&mut self) {
        if let Some(x) = &mut self.0 {
            x.frames.push((Frame::Pop, Instant::now()))
        }
    }

    // We could expose profile on the Heap, but it's an implementation detail that it works here.
    pub(crate) fn gen(&self) -> anyhow::Result<ProfileData> {
        match &self.0 {
            None => Err(FlameProfileError::NotEnabled.into()),
            Some(x) => Ok(Self::gen_profile(x)),
        }
    }

    fn gen_profile(x: &FlameData) -> ProfileData {
        // Need to write out lines which look like:
        // root;calls1;calls2 1
        // All the numbers at the end must be whole numbers (we use milliseconds)
        let mutable_names = x.index.mutable_values.map(|x| x.to_repr());
        let frozen_names = x.index.frozen_values.map(|x| x.to_value().to_repr());
        ProfileData {
            profile_mode: ProfileMode::TimeFlame,
            profile: ProfileDataImpl::TimeFlameProfile(
                Stacks::new(&mutable_names, &frozen_names, &x.frames).render(),
            ),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;
    use std::thread;
    use std::time::Duration;

    use anyhow::Context;
    use starlark_derive::starlark_module;

    use crate as starlark;
    use crate::assert::Assert;
    use crate::environment::Globals;
    use crate::environment::GlobalsBuilder;
    use crate::environment::Module;
    use crate::eval::runtime::file_loader::ReturnOwnedFileLoader;
    use crate::eval::Evaluator;
    use crate::eval::ProfileMode;
    use crate::syntax::AstModule;
    use crate::syntax::Dialect;
    use crate::values::none::NoneType;

    #[test]
    fn test_time_flame_works_inside_frozen_module() {
        #[starlark_module]
        fn register_sleep(globals: &mut GlobalsBuilder) {
            fn sleep() -> anyhow::Result<NoneType> {
                thread::sleep(Duration::from_millis(2));
                Ok(NoneType)
            }
        }

        let mut a = Assert::new();
        a.globals_add(register_sleep);
        let a_bzl = a.pass_module(
            r#"
def foo():
    for i in range(5):
        # Must sleep otherwise time flame will round the duration to zero and erase it.
        sleep()
    "#,
        );

        let modules = HashMap::from_iter([("a.bzl".to_owned(), a_bzl)]);
        let loader = ReturnOwnedFileLoader { modules };

        let module = Module::new();
        let mut eval = Evaluator::new(&module);
        eval.enable_profile(&ProfileMode::TimeFlame).unwrap();
        eval.set_loader(&loader);
        eval.eval_module(
            AstModule::parse(
                "x.star",
                r#"
load("a.bzl", "foo")

def bar():
    for i in range(10):
        foo()

bar()
"#
                .to_owned(),
                &Dialect::Standard,
            )
            .unwrap(),
            &Globals::standard(),
        )
        .unwrap();

        let profile = eval.gen_profile().unwrap().gen().unwrap();
        let the_line = profile
            .lines()
            .find(|l| l.contains("foo"))
            .with_context(|| {
                format!(
                    "There must be a line with `foo` in the profile: {:?}",
                    profile
                )
            })
            .unwrap();
        assert!(
            the_line.contains("bar"),
            "Profile must contain a line `bar.*foo`: {:?}",
            profile
        );
    }
}
