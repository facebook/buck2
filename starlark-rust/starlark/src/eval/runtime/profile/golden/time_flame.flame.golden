# @generated
# To regenerate, run:
# ```
# STARLARK_RUST_REGENERATE_GOLDEN_TESTS=1 cargo test -p starlark --lib
# ```

root 35
root;"module" 70
root;"module";test.star.test 168
root;"module";test.star.test;test.star.inner 140
